ast/util.go

```diff
@@ -0,0 +1,171 @@
+// limitations under the License.
+
+package ast
+
+import (
+	"strings"
+)
+
+// TextElement is an element of a SQLSentence.
+type TextElement struct {
+	// Next is the next element in a SQLSentence.
+	Next *TextElement
+
+	// Next is the previous element in a SQLSentence.
+	Prev *TextElement
+
+	// The text stored with this element.
+	Text string
+}
+
+type sqlSentenceStatus int8
+
+const (
+	sqlSentenceNormal sqlSentenceStatus = iota
+	sqlSentenceNearBrackets
+	sqlSentenceInBrackets
+)
+
+// SQLSentence express a sql text.
+// In order to avoid copying when string splicing occurs,
+// SQLSentence uses a linked list structure to store strings.
+type SQLSentence struct {
+	root   TextElement
+	status sqlSentenceStatus
+}
+
+// NewSQLSentence returns a empty SQLSentence.
+func NewSQLSentence() *SQLSentence {
+	s := new(SQLSentence)
+	s.Reset()
+	return s
+}
+
+// Reset makes the SQLSentence empty.
+func (s *SQLSentence) Reset() {
+	s.root.Next = &s.root
+	s.root.Prev = &s.root
+	s.status = sqlSentenceNormal
+}
+
+// Text will append text to SQLSentence.
+func (s *SQLSentence) Text(text string) *SQLSentence {
+	s.TextAround(text, "")
+	return s
+}
+
+// TextAround will add `around` around `text` then append to SQLSentence.
+// Example: s.TextAround("xx", "tt") will append "ttxxtt" to SQLSentence.
+func (s *SQLSentence) TextAround(text string, around string) *SQLSentence {
+	s.insertComma()
+	s.InsertText(text, s.root.Prev, around)
+	return s
+}
+
+// Space will append " " to SQLSentence.
+func (s *SQLSentence) Space() *SQLSentence {
+	s.InsertText(" ", s.root.Prev, "")
+	return s
+}
+
+// Sentence will append sentence to SQLSentence.
+func (s *SQLSentence) Sentence(sentence *SQLSentence) *SQLSentence {
+	s.SentenceAround(sentence, "")
+	return s
+}
+
+// SentenceAround will add `around` around `sentence` then append to SQLSentence.
+// Just like TextAround
+func (s *SQLSentence) SentenceAround(sentence *SQLSentence, around string) *SQLSentence {
+	s.insertComma()
+	s.InsertSentence(sentence, s.root.Prev, around)
+	return s
+}
+
+// InsertText inserts `text` surrounded by `around` after at
+func (s *SQLSentence) InsertText(text string, at *TextElement, around string) {
+	if len(around) == 0 {
+		s.insertText(text, at)
+		return
+	}
+	s.insertText(around, at)
+	s.insertText(text, at)
+	s.insertText(around, at)
+}
+
+// InsertSentence inserts `sentence` surrounded by `around` after at
+func (s *SQLSentence) InsertSentence(sentence *SQLSentence, at *TextElement, around string) {
+	if len(around) == 0 {
+		s.insertSentence(sentence, at)
+		return
+	}
+	s.insertText(around, at)
+	s.insertSentence(sentence, at)
+	s.insertText(around, at)
+	sentence.Reset()
+}
+
+// LeftBracket append a "(" to SQLSentence,
+// and after that all append operations(Text, TextAround, Sentence, SentenceAround) will automatically append a comma.
+func (s *SQLSentence) LeftBracket() *SQLSentence {
+	s.Text("(")
+	s.status = sqlSentenceNearBrackets
```

Comment:
This is gonna break when there are nested call to `LeftBracket()`. 

```go
ss.LeftBracket().Text("w").Text("x")
    .LeftBracket().Text("a").Text("b").RightBracket()
    .Text("y").Text("z").RightBracket()
// "(w, x, (a, b)yz)"
```

I'm not sure who will use `LeftBracket()` but I don't see much point when this can be easily solved by an `if` in a `for` loop.